stan/math/rev/fun/rep_row_vector.hpp

```diff
@@ -0,0 +1,35 @@
+#ifndef STAN_MATH_REV_FUN_REP_ROW_VECTOR_HPP
+#define STAN_MATH_REV_FUN_REP_ROW_VECTOR_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/prim/err.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Overload for `var_value<Vector>`.
+ * @tparam T_ret The user supplied return type.
+ * @tparam T A double or var type
+ * @param x The type to be propogated through the new vector.
+ * @param n The size of the new vector.
+ */
+template <typename T_ret, typename T, require_var_matrix_t<T_ret>* = nullptr,
+          require_eigen_row_vector_t<value_type_t<T_ret>>* = nullptr,
+          require_stan_scalar_t<T>* = nullptr>
+inline auto rep_row_vector(const T& x, int n) {
+  check_nonnegative("rep_vector", "n", n);
+  return make_callback_var(value_type_t<T_ret>::Constant(n, value_of(x)),
+                           [x](auto& vi) mutable {
+                             if (is_var<T>::value) {
+                               forward_as<var>(x).adj() += vi.adj().sum();
+                             }
```

Comment:
In a stan program you can have 

```stan
data {
  int N;
  real x;
}
transformed parameters {
  vector[N] y = rep_vector(x, 10);
}
```

So in the C++ we want `y = rep_vector<var_value<Eigen::VectorXd>>(x, 10);`

Reply:
That is kinda weird example. I doubt any model exists where it makes sense to do this (this should go into transformed data). If someone does this anyway we can still pass double to a function that accepts var. There will be a small performance penalty, but we get simpler code - both in math and what gets generated by stanc. 

If you really want to avoid that penalty changing stanc to so that it makes y a prim vector (or even moves it into transformed data block) makes more sense.

Reply:
> If someone does this anyway we can still pass double to a function that accepts var.

Actually ... what would happen is that prim version of the `rep_vector` would be called and its result would get converted into a varmat vector. The rest of my response is still correct.

Reply:
> That is kinda weird example. I doubt any model exists where it makes sense to do this (this should go into transformed data). 

It's in the language so we do have to support it. Though yeah it would be much better for someone to do this in transformed data.

> If someone does this anyway we can still pass double to a function that accepts var. There will be a small performance penalty, but we get simpler code - both in math and what gets generated by stanc.

I think this would mess with the compilers `optimize_ad_levels` pass (which does the thing to make y prim a vector). And if we have this return type only work with var then it feels like a thing I'm going to have to code around as I'm writing the explicit return type stuff for these functions in the compiler. I'd like to just write one function in the compiler that handles all of these cases where we have to set the return type to a function, but if this particular function only accepts vars it feels like a thing I'll end up having to code around.

What part of this is odd to you? The template and the if var? that feels like it's pretty inline with how we handle conditionally taking adjoints.

> If you really want to avoid that penalty changing stanc so that it makes y a prim vector (or even moves it into transformed data block) makes more sense.

The compiler can do the thing where it makes `y` a prim vector, though its in the optimizations (`optimize_ad_levels`) which are not turned on by default. We don't have anything yet to move things across blocks though that would also be very nice as that would let these things use the tricks we have in transformed data for not copying data.

Reply:
We supported that in the language even before this PR. The function just returns a prim type, which gets than converted. 

>> I'd like to just write one function in the compiler that handles all of these cases where we have to set the return type to a function, but if this particular function only accepts vars it feels like a thing I'll end up having to code around.

I dont think there is anything to code around. There is no reason for requesting var return type when passing in double or vice versa. I do not understand why we would support that. It is NOT needed by the language - when passing in double you should set prim return type and when passing in var the rev return type. 

If we start supporting arbitrary prim/rev combinations of input and and output types for one function, we should probably do it for every function. Which is a lot of work for making some technically supported use cases nobody will ever use a tiny bit faster. So I think it is a bad idea.

>> What part of this is odd to you?

Function returning a rev type when called with only prim arguments.

>> The compiler can do the thing where it makes y a prim vector, though its in the optimizations (optimize_ad_levels) which are not turned on by default.

Than we really don't need to optimize this in another place, right? If somebody needs these optimizations they can turn them on. And I guess eventually optimizations will be the default setting.

Reply:
Okay so finally able to come back to this, after messing around in the compiler @t4c1 I think your right and think should just take var inputs. I was worried I'd have to do some weird stuff to get this to look right but the way your suggesting makes the most sense as data only would use the prim impl and we don't do an explicit return in that case